src/java/org/apache/cassandra/index/sai/memory/BlockIndex.java

```diff
@@ -0,0 +1,577 @@
+    {
+        return isLiteral ? version -> ByteSourceInverse.unescape(ByteSource.peekable(term.asComparableBytes(version)))
+                         : term;
+    }
+
+    private ByteSource append(ByteSource src, int lastByte)
+    {
+        return new ByteSource()
+        {
+            boolean done = false;
+
+            @Override
+            public int next()
+            {
+                if (done)
+                    return END_OF_STREAM;
+                int n = src.next();
+                if (n != END_OF_STREAM)
+                    return n;
+
+                done = true;
+                return lastByte;
+            }
+        };
+    }
+
+    public static ByteComparable nudge(ByteComparable comp)
+    {
+        byte[] bytes = ByteSourceInverse.readBytes(comp.asComparableBytes(ByteComparable.Version.OSS41));
+        byte[] bytes2 = nudge(bytes);
+        return ByteComparable.fixedLength(bytes2);
+    }
+
+    // append unsigned byte 0 to the array
+    public static byte[] nudge(byte[] orig)
+    {
+        byte[] bytes = Arrays.copyOf(orig, orig.length + 1);
+        bytes[orig.length] = (int) 0; // unsigned requires cast
+        return bytes;
+    }
+
+    public static ByteComparable nudgeReverse(ByteComparable comp)
+    {
+        byte[] bytes = ByteSourceInverse.readBytes(comp.asComparableBytes(ByteComparable.Version.OSS41));
+        byte[] bytes2 = nudgeReverse(bytes);
+        return ByteComparable.fixedLength(bytes2);
+    }
+
+    // decrement the unsigned byte array
+    public static byte[] nudgeReverse(byte[] orig)
+    {
+        byte[] bytes = null;
+
+        // the only unsigned int is 0 so return an empty byte array which is lesser
+        if (orig.length == 1 && UnsignedBytes.toInt(orig[0]) == 0)
+        {
+            return new byte[0];
+        }
+
+        int idx = orig.length - 1;
+
+        bytes = Arrays.copyOf(orig, orig.length);
+
+        while (true)
+        {
+            // iterate backwards until the value isn't 0
+            if (orig[idx] == ((byte) 0) && idx > 0)
+            {
+                byte b = bytes[idx - 1];
+                bytes[idx - 1] = unsignedDecrement(b);
+                idx--;
+            }
+            else
+            {
+                // previous is unsigned 0, remove the element
+                if (idx < orig.length - 1 && orig[idx + 1] == ((byte) 0))
+                {
+                    return Arrays.copyOf(orig, orig.length - 1);
+                }
+
+                // decrement the byte
+                byte b = bytes[bytes.length - 1];
+                bytes[bytes.length - 1] = unsignedDecrement(b);
+                // fill the remainder with a bunch of 255's
+                // 500 may be replaced with the exact term max length
+                byte[] bytes2 = Arrays.copyOf(bytes, bytes.length + 500);
+                Arrays.fill(bytes2, bytes.length, bytes2.length, (byte)255);
+                return bytes2;
+            }
+        }
+    }
```

Comment:
Needlessly complex code. 

Can be rewritten to:
```java
    public static byte[] nudgeReverse(byte[] orig)
    {
        assert orig.length > 0 : "Input byte array must be not empty";
        int lastIdx = orig.length - 1;

        if (UnsignedBytes.toInt(orig[lastIdx]) == 0)
        {
            return Arrays.copyOf(orig, lastIdx);
        }
        else
        {
            byte[] result = Arrays.copyOf(orig, orig.length + 500);
            result[lastIdx] = unsignedDecrement(orig[lastIdx]);
            Arrays.fill(result, orig.length, result.length, (byte)255);
            return result;
        }
    }
```

But that's not the major problem.
The "500" magic is going to cause a very subtle correctness issue if a term in data is longer by more than 500 bytes. 
Something which may be extremely unlikely, but still not exactly impossible.
And bugs that are extremely unlikely, yet still possible, are the ones of the worst kind of bugs - because you can spend a month trying to reproduce.

A few options:
1. Replace 500 with a constant. Add an assert on length of every value added to the index - which must not exceed the length limit. But I don't think it is acceptable to have a hard limit on the term size - this would certainly require a user documentation change and might break some user programs.
2. Track the length of the longest term added so far. This has additional benefit of not overallocating when the terms are shorter than 500 bytes.
3. Drop the whole "nudge" idea and handle inclusiveness by switching the comparison type, not the bounds.

Reply:
There was a comment left in the previous commit addressing this issue.

> // 500 may be replaced with the exact term max length

And it's solved.

Reply:
Looks good.